File: Backend/app/api/routes/ppl.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List, Optional

from DSL_PPL.src.drink_semantics import DrinkPreference, PreferenceVisitor
from DSL_PPL.src.syntax_checker import check_syntax 
from DSL_PPL.src.mongo_query import build_mongo_query
from app.db.database import product_collection


router = APIRouter(
    prefix="/ppl",
    tags=["ppl"],
)


class ParseRequest(BaseModel):
  text: str


class PreferenceOut(BaseModel):
  temperature: Optional[str] = None
  baseType: Optional[str] = None
  sweetness: Optional[str] = None
  caffeine: Optional[str] = None
  size: Optional[str] = None


class RecommendRequest(BaseModel):
  temperature: Optional[str] = None
  baseType: Optional[str] = None
  sweetness: Optional[str] = None
  caffeine: Optional[str] = None
  size: Optional[str] = None


class ProductOut(BaseModel):
  id: str
  name: str
  description: Optional[str] = None
  category: Optional[str] = None        
  drinkCategory: Optional[str] = None   
  temperatures: Optional[List[str]] = None
  sweetnessLevel: Optional[str] = None
  price: Optional[float] = None
  imageURL: Optional[str] = None


@router.post("/parse", response_model=PreferenceOut)
def parse_text(body: ParseRequest):
    ok, errs, tree = check_syntax(body.text)

    if not ok:
        raise HTTPException(
            status_code=400,
            detail={
                "message": (
                    "Syntax Error.\n"
                    "Example:\n"
                    "- \"I want a cold coffee\"\n"
                    "- \"Give me a iced tea with low sugar\""
                )
            },
        )

    visitor = PreferenceVisitor()
    pref: DrinkPreference = visitor.visit(tree) 

    if all(
        getattr(pref, field) is None
        for field in ("temperature", "baseType", "sweetness", "caffeine", "size")
    ):
        raise HTTPException(
            status_code=400,
            detail={
                "message": (
                    "Thiếu thông tin đồ uống.\n"
                    "Bạn cần nói rõ hơn, ví dụ:\n"
                    "- \"I want a cold coffee\"\n"
                    "- \"Give me a warm tea without caffeine\""
                )
            },
        )

    return PreferenceOut(
        temperature=pref.temperature,
        baseType=pref.baseType,
        sweetness=pref.sweetness,
        caffeine=pref.caffeine,
        size=pref.size,
    )


@router.post("/recommend", response_model=List[ProductOut])
async def recommend_products(body: RecommendRequest):
    pref = DrinkPreference(
        temperature=body.temperature,
        baseType=body.baseType,
        sweetness=body.sweetness,
        caffeine=body.caffeine,
        size=body.size,
    )
    if pref.caffeine is not None:
        return []
    
    print("PREF:", pref)
    mongo_filter = build_mongo_query(pref)

    print("MONGO FILTER:", mongo_filter) 

    coll = product_collection()
    docs = await coll.find(mongo_filter).to_list(length=50)

    print("DOCS COUNT:", len(docs))

    products: List[ProductOut] = []
    for doc in docs:
        products.append(
            ProductOut(
                id=str(doc.get("_id")),
                name=doc.get("name", ""),
                description=doc.get("description"),
                category=doc.get("category"),
                drinkCategory=doc.get("drinkCategory"),
                temperatures=doc.get("temperatures"),
                sweetnessLevel=doc.get("sweetnessLevel"),
                price=doc.get("price"),
                imageURL=doc.get("imageURL"),
            )
        )

    return products

File: ecommercial_shopping/lib/core/models/ppl.dart
class PplParseResult {
  final String? temperature;
  final String? baseType;
  final String? sweetness;
  final String? caffeine;
  final String? size;

  const PplParseResult({
    this.temperature,
    this.baseType,
    this.sweetness,
    this.caffeine,
    this.size,
  });

  factory PplParseResult.fromJson(Map<String, dynamic> json) {
    String? normalize(dynamic v) {
      if (v == null) return null;
      if (v is String && v.toLowerCase() == 'null') return null;
      return v.toString();
    }

    return PplParseResult(
      temperature: normalize(json['temperature']),
      baseType: normalize(json['baseType']),
      sweetness: normalize(json['sweetness']),
      caffeine: normalize(json['caffeine']),
      size: normalize(json['size']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'temperature': temperature,
      'baseType': baseType,
      'sweetness': sweetness,
      'caffeine': caffeine,
      'size': size,
    };
  }
}

class PplDrinkRecommendation {
  final String id;
  final String name;
  final String description;
  final String category;
  final String drinkCategory;
  final List<String> temperatures;
  final String? sweetnessLevel;
  final double price;
  final String? imageURL;

  const PplDrinkRecommendation({
    required this.id,
    required this.name,
    required this.description,
    required this.category,
    required this.drinkCategory,
    required this.temperatures,
    this.sweetnessLevel,
    required this.price,
    this.imageURL,
  });

  factory PplDrinkRecommendation.fromJson(Map<String, dynamic> json) {
    return PplDrinkRecommendation(
      id: (json['id'] ?? json['_id'] ?? '') as String,
      name: json['name'] as String? ?? '',
      description: json['description'] as String? ?? '',
      category: json['category'] as String? ?? '',
      drinkCategory: json['drinkCategory'] as String? ?? '',
      temperatures: (json['temperatures'] as List<dynamic>? ?? const [])
          .map((e) => e.toString())
          .toList(),
      sweetnessLevel: json['sweetnessLevel'] as String?,
      price: (json['price'] as num?)?.toDouble() ?? 0,
      imageURL: json['imageURL'] as String?,
    );
  }
}
